core/src/main/scala/chisel3/Layer.scala

```diff
@@ -43,6 +43,11 @@ object layer {
     private[chisel3] def sourceInfo: SourceInfo = _sourceInfo
 
     private[chisel3] def name: String = simpleClassName(this.getClass())
+
+    private[chisel3] lazy val fullName: String = parent match {
```

Comment:
Does this need to be lazy?

Reply:
Looking at it now. No, it doesn't need to be. It's memoized, even if it's eager. The only bad thing to do here would make this a `def`. I'll change it.

Reply:
Actually, this may need to be lazy to work around object initialization order issues. If this is not lazy, I'm seeing:

```
[info] chiselTests.LayerSpec *** ABORTED ***
[info]   java.lang.ExceptionInInitializerError:
[info]   at chisel3.layer$Layer$.<clinit>(Layer.scala:55)
[info]   at chisel3.internal.DynamicContext.<init>(Builder.scala:529)
[info]   at chisel3.stage.phases.Elaborate.$anonfun$transform$1(Elaborate.scala:39)
[info]   at scala.collection.immutable.List.flatMap(List.scala:293)
[info]   at scala.collection.immutable.List.flatMap(List.scala:79)
[info]   at chisel3.stage.phases.Elaborate.transform(Elaborate.scala:28)
[info]   at chisel3.stage.phases.Elaborate.transform(Elaborate.scala:21)
[info]   at firrtl.options.DependencyManager.$anonfun$transform$5(DependencyManager.scala:280)
[info]   at firrtl.Utils$.time(Utils.scala:53)
[info]   at firrtl.options.DependencyManager.$anonfun$transform$3(DependencyManager.scala:280)
[info]   ...
[info]   Cause: java.lang.NullPointerException: Cannot invoke "Object.equals(Object)" because "chisel3.layer$Layer$Root$.MODULE$" is null
[info]   at chisel3.layer$Layer.<init>(Layer.scala:48)
[info]   at chisel3.layer$Layer$Root$.<init>(Layer.scala:54)
[info]   at chisel3.layer$Layer$Root$.<clinit>(Layer.scala:54)
[info]   at chisel3.layer$Layer$.<clinit>(Layer.scala:55)
[info]   at chisel3.internal.DynamicContext.<init>(Builder.scala:529)
[info]   at chisel3.stage.phases.Elaborate.$anonfun$transform$1(Elaborate.scala:39)
[info]   at scala.collection.immutable.List.flatMap(List.scala:293)
[info]   at scala.collection.immutable.List.flatMap(List.scala:79)
[info]   at chisel3.stage.phases.Elaborate.transform(Elaborate.scala:28)
[info]   at chisel3.stage.phases.Elaborate.transform(Elaborate.scala:21)
[info]   ...
```

I can avoid this by doing a match on null. However, that doesn't seem right.

Any ideas?

Reply:
Matching on `null` won't help since you'll probably get the wrong value. Alright just make it lazy and leave a comment that this is why it's lazy.

Reply:
I misread the code here. The problem is that for this to be a non-lazy val, then it has to actually be defined for the root layer. This, for serialization, never actually occurs. However, once I started playing with some error messages for define the issue became obvious. I'll update this to be non-lazy.

Reply:
Fixed in https://github.com/chipsalliance/chisel/pull/3744/commits/780d07f3ab384ff2d5613f324b72e06016ecfe9e.